src/azure-cli-core/azure/cli/core/_profile.py

```diff
@@ -229,7 +235,24 @@ def login_with_managed_identity(self, identity_id=None, allow_no_subscriptions=N
         from azure.cli.core.auth.adal_authentication import MSIAuthenticationWrapper
         resource = self.cli_ctx.cloud.endpoints.active_directory_resource_id
 
-        if identity_id:
+        if not any((identity_id, client_id, object_id, resource_id)):
+            identity_type = MsiAccountTypes.system_assigned
+            msi_creds = MSIAuthenticationWrapper(resource=resource)
+        elif client_id:
+            identity_type = MsiAccountTypes.user_assigned_client_id
+            identity_id = client_id
+            msi_creds = MSIAuthenticationWrapper(resource=resource, client_id=client_id)
+        elif object_id:
+            identity_type = MsiAccountTypes.user_assigned_object_id
+            identity_id = object_id
+            msi_creds = MSIAuthenticationWrapper(resource=resource, object_id=object_id)
+        elif resource_id:
+            identity_type = MsiAccountTypes.user_assigned_resource_id
+            identity_id = resource_id
+            msi_creds = MSIAuthenticationWrapper(resource=resource, msi_res_id=resource_id)
```

Comment:
The ID names are aligned with MSAL: https://learn.microsoft.com/en-us/entra/msal/python/advanced/managed-identity#user-assigned-managed-identities

- Client ID (`client_id`)
- Resource ID (`resource_id`) - inconsistent with `msrestazure`'s `msi_res_id`
- Object ID (`object_id`)